Gems/Atom/RHI/WebGPU/Code/Source/RHI/BufferPool.cpp

```diff
@@ -5,21 +5,189 @@
+    {
+        auto& buffer = static_cast<Buffer&>(resource);
+
+        // Wait for any pending streaming upload.
+        // device.GetAsyncUploadQueue().WaitForUpload(buffer.GetUploadHandle());
+
+        if (auto* resolver = GetResolver())
+        {
+            ResourcePoolResolver* poolResolver = static_cast<ResourcePoolResolver*>(resolver);
+            poolResolver->OnResourceShutdown(resource);
+        }
+
+        RHI::HeapMemoryLevel heapMemoryLevel = GetDescriptor().m_heapMemoryLevel;
+        RHI::HeapMemoryUsage& heapMemoryUsage = m_memoryUsage.GetHeapMemoryUsage(heapMemoryLevel);
+        size_t sizeInBytes = buffer.GetDescriptor().m_byteCount;
+        heapMemoryUsage.m_usedResidentInBytes -= sizeInBytes;
+        heapMemoryUsage.m_totalResidentInBytes -= sizeInBytes;
+
+        // Deallocate the BufferMemory
+        buffer.Invalidate();
+    }
+
+    RHI::ResultCode BufferPool::OrphanBufferInternal(RHI::DeviceBuffer& bufferBase)
+    {
+        auto& buffer = static_cast<Buffer&>(bufferBase);
+
+        // Deallocate the BufferMemory
+        buffer.Invalidate();
+
+        auto result = InitBufferInternal(bufferBase, bufferBase.GetDescriptor());
+        RETURN_RESULT_IF_UNSUCCESSFUL(result);
+
+        buffer.InvalidateViews();
+        return RHI::ResultCode::Success;
+    }
+
+    RHI::ResultCode BufferPool::MapBufferInternal(const RHI::DeviceBufferMapRequest& mapRequest, RHI::DeviceBufferMapResponse& response)
+    {
+        const RHI::BufferPoolDescriptor& descriptor = GetDescriptor();
+        auto* buffer = static_cast<Buffer*>(mapRequest.m_buffer);
+
+        void* mappedData = nullptr;
+        auto& instance = Instance::GetInstance().GetNativeInstance();
+        if (buffer->CanBeMap())
+        {
+            size_t byteOffset = RHI::AlignDown(mapRequest.m_byteOffset, MapOffsetAligment);
+            size_t byteCount = RHI::AlignUp(mapRequest.m_byteCount, MapSizeAligment);
+            instance.WaitAny(
+                buffer->GetNativeBuffer().MapAsync(
+                    ConvertMapMode(descriptor.m_hostMemoryAccess),
+                    byteOffset,
+                    byteCount,
+                    wgpu::CallbackMode::WaitAnyOnly,
+                    [&](wgpu::MapAsyncStatus status, const char* message)
+                    {
+                        if (status != wgpu::MapAsyncStatus::Success)
+                        {
+                            AZ_Assert(false, "Failed to map buffer: %s", message);
+                            return;
+                        }
+                        mappedData = buffer->GetNativeBuffer().GetMappedRange(byteOffset, byteCount);
+                    }),
+                UINT64_MAX);
+
+            if (mappedData)
+            {
+                mappedData = static_cast<int8_t*>(mappedData) + mapRequest.m_byteOffset - byteOffset;
+            }
+            else
+            {
+                return RHI::ResultCode::Fail;
+            }
+        }
+        else
```

Comment:
maybe put the mappedData inside of the if, this way you can put it into the else and no need further nesting